### ISSUE TITLE
Sync body language and eye contact in 2D mode (via eye and facial expression tracking)

### ISSUE THREAD

**Issue by djay** [mozilla/hubs#3689]
**Is your feature request related to a problem? Please describe.**
Eye contact and other non verbal language is very important to hold smooth conversations. Spatial audio helps some non verbal queues would improve the quality of communication enormously IMO. If someone looks confused then it should be obvious on their avatar. If everyone is looking toward a certain avatar it should be obvious to them they have the groups attention.
This video gives a good video demo - https://blog.tobii.com/why-eye-tracking-will-be-standard-on-all-vr-headsets-f1b0e5326f3c

**Describe the solution you'd like**
For desktop and mobile users it should be possible to do eye tracking (at least rough eye tracking), and facial recognition in javascript. Headsets will catch up with hardware solutions later. This information is streamed and then represented on the avatar (avatars might need to be more flexible?).

Even if eye tracking could not be done for headset user without special hardware, facial expression tracking could still be done if the user is sitting in front of a laptop or phone camera.

**Describe alternatives you've considered**
Streaming video of a users face - https://github.com/mozilla/hubs/issues/2327, https://github.com/mozilla/hubs/issues/3228 This requires more bandwidth and it also doesn't solve the eye contact problem as video cameras are offset from where the user is looking.

https://github.com/mozilla/hubs/issues/3285 is more about triggering avatar emotions manually.

https://github.com/mozilla/hubs/issues/2256 is about turning the head instead of the whole body in 2D mode which would work well in conjunction with this eye tracking enhancement. It might even be possible to use eye tracking to turn the head when the eyes point off the screen. It should also be possible to do head orientation tracking using a webcam and sync this with an avatar.

**Additional context**

**Comment by djay** [comment]
This could work for syncing facial expressions - https://github.com/tensorflow/tfjs-models/tree/master/facemesh. not sure if facial expressions would need to be further reduced to smile, frown etc or the mesh used to changed other avatars. but how to map to non humans?

**Comment by luser** [comment]
@lazerwalker built a demo of something similar using 2D avatars:
https://github.com/lazerwalker/spatial-webrtc-test

**Comment by lazerwalker** [comment]
Yep! That project is based on https://github.com/yemount/pose-animator, refactored to separate out face/body tracking from avatar rendering so it can send joint positions over a WebRTC connection and render remotely. It's using the aforementioned facemesh for facial landmarking, and [posenet](https://github.com/tensorflow/tfjs-models/tree/master/posenet) for body tracking.

An important thing to note is it would be _far_ easier to do this as a flat 2D face texture rather than morphing a full 3D face model. WebGL's severe limitations on the number of morph targets makes it infeasible to do full facial expression tracking on a 3D model rather than reducing to simple predetermined animations/expressions (as @djay mentions) barring clever 3D rendering tricks that are beyond my paygrade. Conversely, I suspect it'd be relatively doable to render a 2D avatar face as in those demos and just slap that texture onto a head model with a flat face surface. Using pose tracking for head and arm positions seems eminently doable with the existing 3D models.

**Comment by djay** [comment]
@lazerwalker that demo is super cool. You can already see how much emotion it could add to a conversation.
I agree 2d texture is probably enough. probably even a feed of smile/frown/suprise etc would add a lot (plus eye gaze and head orientation). Apparently [humans only have 21](https://www.cbsnews.com/news/computer-program-spots-21-distinct-facial-expressions/). 

I notice that demo doesn't do eye gaze, which I suspect is very important. Although [some studies show eye gaze doesn't work the same in VR](https://www.tuni.fi/en/news/eye-contact-different-virtual-reality-and-real-life) I think that depends on how realistic the avatar is and if its about meassuring emotional connection vs conversation turn taking. I'm not sure on [research on turn taking but this article discusses the issues well](https://medium.com/@marindabotha/eye-contact-in-virtual-reality-e6ac72d4fc45).
I suspect the eye tracking needed would not have to be that accurate. Not nearly as accurate as the systems needed for controlling a cursor. Maybe only 15-20deg or more accuracy? I suspect facemesh is capable doing that already.
But imagine feeling eyes following you when you enter a group to join a conversation vs no one noticing you.

Pose tracking and arm tracking would certainly help. but if I think if most users are sitting at laptops they might not often have hands or arms visibile?

**Comment by djay** [comment]
In addition to turn taking and body language just knowing if someone really there or not would be a big help from face detection. In a recent meetup https://github.com/mozilla/hubs/discussions/3718 some people had to step away and you might try to interact with someone not currently in front of their laptop. Perhaps such an avatar could be represented as sleeping?

**Comment by djay** [comment]
@lazerwalker do you know if anyone at mozilla are considering building something like this this webrtc sending of facial expressions into hubs?

**Comment by djay** [comment]
https://uploadvr.com/zuckerberg-quest-3-4-eye-face-tracking/
"When first presenting codec avatars, Facebook warned the technology was still “years away” for consumer products – shipping this kind of photoreal avatar will require a number of breakthroughs."
